File: java/QuickSort.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * A class containing the implementation of the Quick Sort algorithm.
 */
public class QuickSort {

    /**
     * The main method that sorts an array using the quick sort algorithm.
     *
     * @param arr The array to be sorted.
     */
    public void sort(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        quickSort(arr, 0, arr.length - 1);
    }

    /**
     * A recursive helper method to sort the array.
     *
     * @param arr  The array to be sorted.
     * @param low  The starting index.
     * @param high The ending index.
     */
    private void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // pi is partitioning index, arr[pi] is now at right place
            int pi = partition(arr, low, high);

            // Recursively sort elements before partition and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    /**
     * This method takes the last element as pivot, places the pivot element at its
     * correct position in the sorted array, and places all smaller (smaller than pivot)
     * elements to the left of the pivot and all greater elements to the right of the pivot.
     *
     * @param arr  The array to be partitioned.
     * @param low  The starting index.
     * @param high The ending index.
     * @return The partitioning index.
     */
    private int partition(int[] arr, int low, int high) {
        // Choose the rightmost element as pivot
        int pivot = arr[high];

        // Index of smaller element
        int i = (low - 1);

        for (int j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (arr[j] <= pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i+1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    /**
     * The main method to demonstrate the quick sort functionality interactively.
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        QuickSort sorter = new QuickSort();

        System.out.println("--- Interactive Quick Sort ---");
        System.out.print("Enter the number of elements in the array: ");
        int n = scanner.nextInt();

        if (n <= 0) {
            System.out.println("Array size must be positive. Exiting.");
            scanner.close();
            return;
        }

        int[] arr = new int[n];
        System.out.println("Enter the elements of the array:");
        for (int i = 0; i < n; i++) {
            System.out.print("Element " + (i + 1) + ": ");
            arr[i] = scanner.nextInt();
        }

        System.out.println("\nOriginal array: " + Arrays.toString(arr));

        sorter.sort(arr);

        System.out.println("Sorted array:   " + Arrays.toString(arr));

        scanner.close();
    }
}

File: python/quick_sort.py
def quick_sort(arr):
    """
    Quick sort algorithm implementation
    Time Complexity: O(n log n) average, O(n²) worst case
    Space Complexity: O(log n)
    """
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    
    return quick_sort(left) + middle + quick_sort(right)

def quick_sort_inplace(arr, low=0, high=None):
    """
    In-place quick sort implementation
    """
    if high is None:
        high = len(arr) - 1
    
    if low < high:
        # Partition the array and get pivot index
        pivot_index = partition(arr, low, high)
        
        # Sort elements before and after partition
        quick_sort_inplace(arr, low, pivot_index - 1)
        quick_sort_inplace(arr, pivot_index + 1, high)

def partition(arr, low, high):
    """
    Partition function for quick sort
    Uses last element as pivot
    """
    pivot = arr[high]
    i = low - 1  # Index of smaller element
    
    for j in range(low, high):
        # If current element is smaller than or equal to pivot
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort_random_pivot(arr, low=0, high=None):
    """
    Quick sort with random pivot selection
    Better average case performance
    """
    import random
    
    if high is None:
        high = len(arr) - 1
    
    if low < high:
        # Choose random pivot and swap with last element
        random_index = random.randint(low, high)
        arr[random_index], arr[high] = arr[high], arr[random_index]
        
        pivot_index = partition(arr, low, high)
        quick_sort_random_pivot(arr, low, pivot_index - 1)
        quick_sort_random_pivot(arr, pivot_index + 1, high)

def quick_select(arr, k):
    """
    Quick select algorithm to find kth smallest element
    Time Complexity: O(n) average case
    """
    if not arr:
        return None
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    
    if k < len(left):
        return quick_select(left, k)
    elif k < len(left) + len(middle):
        return pivot
    else:
        return quick_select(right, k - len(left) - len(middle))

# Test the functions
if __name__ == "__main__":
    # Test quick sort
    numbers = [64, 34, 25, 12, 22, 11, 90]
    print(f"Original array: {numbers}")
    
    sorted_numbers = quick_sort(numbers.copy())
    print(f"Quick sorted: {sorted_numbers}")
    
    # Test in-place quick sort
    inplace_numbers = numbers.copy()
    quick_sort_inplace(inplace_numbers)
    print(f"In-place quick sorted: {inplace_numbers}")
    
    # Test random pivot quick sort
    random_numbers = numbers.copy()
    quick_sort_random_pivot(random_numbers)
    print(f"Random pivot sorted: {random_numbers}")
    
    # Test quick select
    test_array = [3, 6, 8, 10, 1, 2, 1]
    k = 3
    kth_smallest = quick_select(test_array, k - 1)  # k-1 for 0-based indexing
    print(f"\n{k}rd smallest element in {test_array}: {kth_smallest}")
    
    # Find median using quick select
    sorted_test = sorted(test_array)
    median_index = len(test_array) // 2
    median = quick_select(test_array, median_index)
    print(f"Median of {sorted_test}: {median}")
